### PR TITLE
[WIP] Fix custom op when used with need_top_grad=False

### PR DESCRIPTION
**Pull request by anirudh2290**
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fix custom op when used with need_top_grad=False (and when applicable, API doc)

## Comments ##
- Before this PR, the custom op failed with need_top_grad=False as the backward infer storage logic expected the out_grad to always be present. Requires #8721 to be merged.

### CODE REVIEW

File: python/mxnet/operator.py [hunk 1]
```diff
@@ -563,7 +563,7 @@ def infer_storage_type_backward(self, in_stype):
             list of aux stypes calculated from in_stype,
             in the same order as declared in list_auxiliary_states.
         """
-        return in_stype, [in_stype[0]]*len(self.list_outputs()), \
+        return in_stype, [in_stype[0]]*len(self.list_arguments()), \
```
Comment (piiswrong) [inline python/mxnet/operator.py]: This looks wrong. Why is there only one argument to infer_storage_type_backward?

Comment (anirudh2290) [reply python/mxnet/operator.py]: the argument in_stype is a list of storage types for out_grad, in_data, out_data.

Comment (piiswrong) [reply python/mxnet/operator.py]: That's wrong. There should be multiple lists.

**Comment by piiswrong** [comment]
I overlooked this when merging. But the interface for infer_storage_backward is really bad.
We need to consider reverting the inferstorage PR and redesign the interface

**Pull request closed without merging**